### PR TITLE
Bring harmony to the test snapshot filtering situation

### PR DESCRIPTION
**Pull request by zanieb**
The snapshot filtering situation has gotten way out of hand, with each test hand-rolling it's own filters on top of copied cruft from previous tests.

I've attempted to address this holistically:

- `TestContext.filters()` has everything you should need 
    - This was introduced a while ago, but needed a few more filters for it to be generalized everywhere
    - Using `INSTA_FILTERS` is **not recommended** unless you do not want the context filters
    - It is okay to extend these filters for things unrelated to paths
    - If you have to write a custom path filter, please highlight it in review so we can address it in the common module
- `TestContext.site_packages()` gives cross-platform access to the site-packages directory
    - Do not manually construct the path to site-packages from the venv
    - Do not turn off tests on Windows because you manually constructed a Unix path to site-packages
- `TestContext.workspace_root` gives access to uv's repository directory
    - Use this for installing from `scripts/packages/`
    - If you need coverage for relative paths, copy the test package into the `temp_dir` don't change the working directory of the test fixture

There is additional work that can be done here, such as:

- Auditing and removing additional uses of `INSTA_FILTERS`
- Updating manual construction of `Command` instances to use a utility
- The `venv` tests are particularly frightening in their lack of a test context and could use some love
- Improving the developer experience i.e. apply context filters to snapshots by default

### CODE REVIEW

File: crates/uv-fs/src/path.rs [hunk 1]
```diff
@@ -34,7 +39,9 @@ impl<T: AsRef<Path>> Simplified for T {
 
     fn user_display(&self) -> std::path::Display {
         let path = dunce::simplified(self.as_ref());
-        path.strip_prefix(&*CWD).unwrap_or(path).display()
+        path.strip_prefix(CWD.simplified())
+            .unwrap_or(path)
+            .display()
     }
 }
```
Comment (zanieb) [inline crates/uv-fs/src/path.rs:46]: This fixes a bug where sometimes the path was not properly relativized. This is the only user-facing change here and could be split into a separate pull request. It's needed for consistent cross-platform snapshot replacements.

File: crates/uv/tests/cache_prune.rs [hunk 2]
```diff
@@ -102,25 +97,14 @@ fn prune_stale_directory() -> Result<()> {
...
+    uv_snapshot!(context.filters(), prune_command(&context).arg("--verbose"), @r###"
     success: true
     exit_code: 0
     ----- stdout -----
 
     ----- stderr -----
-    Pruning cache at: [CACHE_DIR]
-    DEBUG Pruning cache at: [CACHE_DIR]/simple-v4
+    Pruning cache at: [CACHE_DIR]/
+    DEBUG Removing dangling cache entry: [CACHE_DIR]/simple-v4
```
Comment (zanieb) [inline crates/uv/tests/cache_prune.rs:107]: This is a great example of a non-standard filter resulting in an incorrect snapshot.

File: crates/uv/tests/cache_prune.rs [hunk 3]
```diff
@@ -145,25 +129,26 @@ fn prune_stale_symlink() -> Result<()> {
...
+        .filters()
+        .into_iter()
+        .chain([
+            // The cache entry does not have a stable key, so we filter it out
+            (
+                r"\[CACHE_DIR\](\\|\/)(.+)(\\|\/).*",
+                "[CACHE_DIR]/$2/[ENTRY]",
+            ),
+        ])
+        .collect();
```
Comment (zanieb) [inline crates/uv/tests/cache_prune.rs:142]: This is one of the two recommended ways to extend the standard filters. I'd generally recommend this. Alternatively, you may want to perform filtering _before_ the context filters in which case you'd just swap the chain.

File: crates/uv/tests/common/mod.rs [hunk 4]
```diff
@@ -56,29 +57,75 @@ impl TestContext {
...
                 .into_iter()
                 .map(|pattern| (pattern, "[CACHE_DIR]/".to_string())),
         );
         filters.extend(
-            Self::path_patterns(&temp_dir)
+            Self::path_patterns(&site_packages)
                 .into_iter()
-                .map(|pattern| (pattern, "[TEMP_DIR]/".to_string())),
+                .map(|pattern| (pattern, "[SITE_PACKAGES]/".to_string())),
         );
```
Comment (zanieb) [inline crates/uv/tests/common/mod.rs:78]: Added a `SITE_PACKAGES` filter which is helpful for cross-platform tests with package install paths in them.

File: crates/uv/tests/common/mod.rs [hunk 5]
```diff
@@ -56,29 +57,75 @@ impl TestContext {
...
         filters.extend(
             Self::path_patterns(&venv)
                 .into_iter()
                 .map(|pattern| (pattern, "[VENV]/".to_string())),
         );
+        filters.extend(
+            Self::path_patterns(&temp_dir)
+                .into_iter()
+                .map(|pattern| (pattern, "[TEMP_DIR]/".to_string())),
+        );
```
Comment (zanieb) [inline crates/uv/tests/common/mod.rs:88]: Moved the `TEMP_DIR` filter down — we want to apply this _after_ all of its subdirectories.

File: crates/uv/tests/common/mod.rs [hunk 6]
```diff
@@ -56,29 +57,75 @@ impl TestContext {
...
+        filters.extend(
+            Self::path_patterns(&workspace_root)
+                .into_iter()
+                .map(|pattern| (pattern, "[WORKSPACE]/".to_string())),
+        );
```
Comment (zanieb) [inline crates/uv/tests/common/mod.rs:93]: Added a filter for the repository workspace — useful for our test packages.

File: crates/uv/tests/common/mod.rs [hunk 7]
```diff
@@ -56,29 +57,75 @@ impl TestContext {
...
+            ),
+            "[SITE_PACKAGES]/".to_string(),
+        ));
+        filters.push((
+            Self::path_pattern(
+                venv.strip_prefix(&temp_dir)
+                    .expect("The test virtual environment directory is always in the tempdir"),
+            ),
+            "[VENV]/".to_string(),
+        ));
```
Comment (zanieb) [inline crates/uv/tests/common/mod.rs:110]: These new filters account for the changes in #2559 — the test commands should always be run with `temp_dir` as the current directory.

File: crates/uv/tests/common/mod.rs [hunk 8]
```diff
@@ -56,29 +57,75 @@ impl TestContext {
...
+
+        // Filter non-deterministic temporary directory names
+        // Note we apply this _after_ all the full paths to avoid breaking their matching
+        filters.push((r"(\\|\/)\.tmp.*(\\|\/)".to_string(), "/[TMP]/".to_string()));
```
Comment (zanieb) [inline crates/uv/tests/common/mod.rs:114]: This addresses cases where a temporary directory with a non-deterministic name is created inside another directory e.g. for package builds we create a temporary directory in our cache.

File: crates/uv/tests/common/mod.rs [hunk 9]
```diff
@@ -128,33 +175,35 @@ impl TestContext {
...
+        let mut patterns = Vec::new();
+
+        // We can only canonicalize paths that exist already
+        if path.as_ref().exists() {
+            patterns.push(Self::path_pattern(
+                path.as_ref()
+                    .canonicalize()
+                    .expect("Failed to create canonical path"),
+            ));
+        }
```
Comment (zanieb) [inline crates/uv/tests/common/mod.rs:189]: This needs to be conditional because sometimes the site packages directory does not exist.

File: crates/uv/tests/common/mod.rs [hunk 10]
```diff
@@ -253,8 +306,8 @@ pub fn bootstrapped_pythons() -> Option<Vec<PathBuf>> {
 
 /// Create a virtual environment named `.venv` in a temporary directory with the given
 /// Python version. Expected format for `python` is "python<version>".
-pub fn create_venv(
-    temp_dir: &assert_fs::TempDir,
+pub fn create_venv<Parent: assert_fs::prelude::PathChild + AsRef<std::path::Path>>(
+    temp_dir: &Parent,
```
Comment (zanieb) [inline crates/uv/tests/common/mod.rs:310]: This is generic because some tests use a child directory to create a second virtual environment. I'm not excited about the approach in those tests, but we can address that later.

File: crates/uv/tests/pip_compile.rs [hunk 11]
```diff
@@ -629,13 +629,7 @@ dependencies = [
 "#,
     )?;
 
-    // In addition to the standard filters, remove the temporary directory from the snapshot.
-    let filters: Vec<_> = [(r"file://.*/", "file://[TEMP_DIR]/")]
-        .into_iter()
-        .chain(INSTA_FILTERS.to_vec())
-        .collect();
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:636]: Filters like these are entirely unnecessary and were handled by `context.filters()` before any changes here, but I believe it wasn't clear that `context.filters()` existed.

File: crates/uv/tests/pip_compile.rs [hunk 12]
```diff
@@ -2247,7 +2228,7 @@ fn compile_wheel_path_dependency() -> Result<()> {
         # via flask
     click==8.1.7
         # via flask
-    flask @ /[TEMP_DIR]/
+    flask @ [TEMP_DIR]/flask-3.0.0-py3-none-any.whl
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:2231]: We don't replace the entire path anymore, just the non-deterministic part. This should generally improve test coverage and clarity.

File: crates/uv/tests/pip_compile.rs [hunk 13]
```diff
@@ -5468,7 +5359,8 @@ fn requires_python_editable() -> Result<()> {
     let context = TestContext::new("3.12");
 
     // Create an editable package with a `Requires-Python` constraint that is not met.
-    let editable_dir = TempDir::new()?;
+    let editable_dir = context.temp_dir.child("editable");
+    editable_dir.create_dir_all()?;
```
Comment (zanieb) [inline crates/uv/tests/pip_compile.rs:5363]: Don't create new temporary directories outside of the context.

File: crates/uv/tests/pip_freeze.rs [hunk 14]
```diff
@@ -136,8 +117,8 @@ fn freeze_duplicate() -> Result<()> {
 
     ----- stderr -----
     warning: The package `pip` has multiple installed distributions:
-    /pip-21.3.1.dist-info
-    /pip-22.1.1.dist-info
+      - [SITE_PACKAGES]/pip-21.3.1.dist-info
+      - [SITE_PACKAGES]/pip-22.1.1.dist-info
```
Comment (zanieb) [inline crates/uv/tests/pip_freeze.rs:121]: Here the non-standard filters clobbered the leading white space.

File: crates/uv/tests/pip_install.rs [hunk 15]
```diff
@@ -596,7 +591,6 @@ fn reinstall_extras() -> Result<()> {
 
 /// Warn, but don't fail, when uninstalling incomplete packages.
 #[test]
-#[cfg(unix)]
```
Comment (zanieb) [inline crates/uv/tests/pip_install.rs:599]: This test was disabled on Windows because the site-packages path was manually constructed.

File: crates/uv/tests/pip_install.rs [hunk 16]
```diff
@@ -756,24 +726,15 @@ fn install_editable() -> Result<()> {
...
         .arg("-e")
-        .arg("../../scripts/packages/poetry_editable")
-        .arg("--strict")
-        .arg("--cache-dir")
-        .arg(context.cache_dir.path())
-        .arg("--exclude-newer")
-        .arg(EXCLUDE_NEWER)
-        .env("VIRTUAL_ENV", context.venv.as_os_str())
-        .env("CARGO_TARGET_DIR", "../../../target/target_install_editable"), @r###"
+        .arg(context.workspace_root.join("scripts/packages/poetry_editable")), @r###"
```
Comment (zanieb) [inline crates/uv/tests/pip_install.rs:737]: There are various instances where the `command` utility was not being used — I'm not sure why but where I see it I've addressed it. Another pass will probably be needed to fix remaining cases.

File: crates/uv/tests/pip_install.rs [hunk 17]
```diff
@@ -859,11 +795,9 @@ fn install_editable_and_registry() -> Result<()> {
     );
 
     // Install the editable version of Black. This should remove the registry-based version.
-    uv_snapshot!(filters, command(&context)
+    uv_snapshot!(context.filters(), command(&context)
         .arg("-e")
-        .arg("../../scripts/packages/black_editable")
-        .current_dir(&current_dir)
-        .env("CARGO_TARGET_DIR", "../../../target/target_install_editable"), @r###"
```
Comment (zanieb) [inline crates/uv/tests/pip_install.rs:866]: Lots of tests set the `CARGO_TARGET_DIR`, presumably for editables that used to use Maturin? I'm actually quite confused but I've removed it from many test cases as it appears superfluous.

File: crates/uv/tests/pip_install_scenarios.rs [hunk 18]
```diff
@@ -2674,7 +2674,7 @@ fn requires_package_prerelease_and_final_any() {
     let context = TestContext::new("3.8");
 
     // In addition to the standard filters, swap out package names for shorter messages
-    let mut filters = INSTA_FILTERS.to_vec();
+    let mut filters = context.filters();
```
Comment (zanieb) [inline crates/uv/tests/pip_install_scenarios.rs:2677]: This change doesn't really matter in the scenario files, but for consistency I've updated the templates.

File: crates/uv/tests/pip_show.rs [hunk 19]
```diff
@@ -85,14 +85,7 @@ fn show_requires_multiple() -> Result<()> {
     );
 
     context.assert_command("import requests").success();
-    let filters = [(
-        r"Location:.*site-packages",
-        "Location: [WORKSPACE_DIR]/site-packages",
-    )]
-    .to_vec();
```
Comment (zanieb) [inline crates/uv/tests/pip_show.rs:92]: We should avoid filtering based on other context in the output (i.e. "Location: ") unless absolutely necessary.

File: crates/uv/tests/pip_list.rs [hunk 20]
```diff
@@ -131,92 +118,47 @@ fn list_editable() -> Result<()> {
...
-        "Editable project location{0}",
-        " ".repeat(workspace_len_difference)
-    );
-    let replace_header = format!("Editable project location{0}", " ".repeat(23));
-
-    let find_whitespace = " ".repeat(25 + workspace_len_difference);
-    let replace_whitespace = " ".repeat(48);
-
-    let search_workspace = workspace_dir_re.as_str().strip_prefix(prefix).unwrap();
-    let replace_workspace = "[WORKSPACE_DIR]/";
```
Comment (zanieb) [inline crates/uv/tests/pip_list.rs:163]: This was just way too complicated and brittle for me. I dropped all of this in favor of some filters that do not preserve the aligned snapshot.

File: crates/uv/tests/pip_list.rs [hunk 21]
```diff
@@ -449,41 +329,25 @@ fn list_exclude() -> Result<()> {
     exit_code: 0
     ----- stdout -----
     Package Version
-    ------- -------
-    anyio   4.3.0
-    idna    3.6
+    [UNDERLINE]
+    anyio 4.3.0
+    idna 3.6
     sniffio 1.3.1
```
Comment (zanieb) [inline crates/uv/tests/pip_list.rs:335]: e.g. this is a worse snapshot, but it's only one line of filters to maintain and I'm not sure if the old snapshot really tested for alignment when it was doing all those manipulations.

**Review by konstin — APPROVED** [review]
:pray:

**Review by BurntSushi — APPROVED** [review]
